File: examples/nirfmx/cdma2k/csharp/VS2010/RFmxCdma2kChp/vb/RFmxCdma2kChp.vb
'Steps:
'1. Open a new RFmx session
'2. Configure the basic instrument properties: Clock Source and Clock Frequency
'3. Configure the basic signal properties: Center Frequency, Reference Level and External Attenuation
'4. Configure the trigger properties
'5. Select CHP measurement and enable the traces
'6. Configure Sweep Time for the CHP measurement
'7. Configure Averaging Parameters for the CHP measurement
'8. Initiate Measurement
'9. Fetch CHP Measurements and Traces
'10. Close the RFmx session

Imports NationalInstruments.RFmx.InstrMX
Imports NationalInstruments.RFmx.Cdma2kMX

Class RFmxCdma2kChp
	Private instrSession As RFmxInstrMX
	Private cdma2k As RFmxCdma2kMX
	Private resourceName As String, frequencySource As String
	Private centerFrequency As Double, referenceLevel As Double, externalAttenuation As Double, frequency As Double, sweepTimeInterval As Double
	Private sweepTimeAuto As RFmxCdma2kMXChpSweepTimeAuto
	Private averagingEnabled As RFmxCdma2kMXChpAveragingEnabled
	Private averagingCount As Integer
	Private averagingType As RFmxCdma2kMXChpAveragingType

	Private digitalEdgeTriggerSource As String
	Private digitalEdgeTriggerEdge As RFmxCdma2kMXDigitalEdgeTriggerEdge
	Private triggerDelay As Double
	Private enableTrigger As Boolean

	Private carrierAbsolutePower As Double, timeout As Double

	Friend Sub Run()
		Try
			InitializeVariables()
			InitializeInstr()
			ConfigureCdma2k()
			RetrieveResults()

			PrintResults()
		Catch ex As Exception
			DisplayError(ex)
		Finally
			' Close session 

			CloseSession()
			Console.WriteLine("Press any key to exit.....")
			Console.ReadKey()
		End Try
	End Sub

	Private Sub InitializeVariables()
		' Initialize input variables 


		resourceName = "RFSA"
		centerFrequency = 833490000.0
		' Hz 
		referenceLevel = 0.0
		' dBm 
		externalAttenuation = 0.0
		' dB 
		timeout = 10.0
		' seconds 

		frequencySource = RFmxInstrMXConstants.OnboardClock
		frequency = 10000000.0
		' Hz 

		'Sweep time
		sweepTimeAuto = RFmxCdma2kMXChpSweepTimeAuto.[True]
		sweepTimeInterval = 0.00167
		' seconds 

		'Averaging
		averagingEnabled = RFmxCdma2kMXChpAveragingEnabled.[False]
		averagingCount = 10
		averagingType = RFmxCdma2kMXChpAveragingType.Rms

		triggerDelay = 0.0
		' seconds 

		digitalEdgeTriggerSource = RFmxInstrMXConstants.Pfi0
		digitalEdgeTriggerEdge = RFmxCdma2kMXDigitalEdgeTriggerEdge.Rising
		enableTrigger = False
	End Sub

	Private Sub InitializeInstr()
		' Create a new RFmx Session 

		instrSession = New RFmxInstrMX(resourceName, "")
	End Sub

	Private Sub ConfigureCdma2k()
		' Get SpecAn signal 

		cdma2k = instrSession.GetCdma2kSignalConfiguration()

		' Configure measurement 

		instrSession.ConfigureFrequencyReference("", frequencySource, frequency)
		cdma2k.ConfigureRF("", centerFrequency, referenceLevel, externalAttenuation)
		cdma2k.ConfigureDigitalEdgeTrigger("", digitalEdgeTriggerSource, digitalEdgeTriggerEdge, triggerDelay, enableTrigger)

		cdma2k.SelectMeasurements("", RFmxCdma2kMXMeasurementTypes.Chp, True)

		cdma2k.Chp.Configuration.ConfigureSweepTime("", sweepTimeAuto, sweepTimeInterval)
		cdma2k.Chp.Configuration.ConfigureAveraging("", averagingEnabled, averagingCount, averagingType)

		cdma2k.Initiate("", "")
	End Sub

	Private Sub RetrieveResults()
		' Retrieve results 

		Dim spectrum As Spectrum(Of Single) = Nothing
		cdma2k.Chp.Results.FetchCarrierAbsolutePower("", timeout, carrierAbsolutePower)
		cdma2k.Chp.Results.FetchSpectrum("", timeout, spectrum)
	End Sub

	Private Sub PrintResults()
		Console.WriteLine("Carrier Absolute Power (dBm)  : {0}", carrierAbsolutePower)
	End Sub

	Private Sub CloseSession()
		Try
			If cdma2k IsNot Nothing Then
				cdma2k.Dispose()
				cdma2k = Nothing
			End If

			If instrSession IsNot Nothing Then
				instrSession.Close()
				instrSession = Nothing
			End If
		Catch ex As Exception
			DisplayError(ex)
		End Try
	End Sub

	Private Shared Sub DisplayError(ex As Exception)
		Console.WriteLine("ERROR:" & vbLf & Convert.ToString(ex.[GetType]()) & ": " & ex.Message)
	End Sub
End Class

File: examples/nirfmx/cdma2k/csharp/VS2010/RFmxCdma2kChp/cs/RFmxCdma2kChp.cs
//Steps:
//1. Open a new RFmx session
//2. Configure the basic instrument properties: Clock Source and Clock Frequency
//3. Configure the basic signal properties: Center Frequency, Reference Level and External Attenuation
//4. Configure the trigger properties
//5. Select CHP measurement and enable the traces
//6. Configure Sweep Time for the CHP measurement
//7. Configure Averaging Parameters for the CHP measurement
//8. Initiate Measurement
//9. Fetch CHP Measurements and Traces
//10. Close the RFmx session

using System;
using NationalInstruments.RFmx.InstrMX;
using NationalInstruments.RFmx.Cdma2kMX;

namespace NationalInstruments.Examples.RFmxCdma2kChp
{
    class RFmxCdma2kChp
    {
        RFmxInstrMX instrSession;
        RFmxCdma2kMX cdma2k;
        string resourceName, frequencySource;
        double centerFrequency, referenceLevel, externalAttenuation, frequency, sweepTimeInterval;
        RFmxCdma2kMXChpSweepTimeAuto sweepTimeAuto;
        RFmxCdma2kMXChpAveragingEnabled averagingEnabled;
        int averagingCount;
        RFmxCdma2kMXChpAveragingType averagingType;

        string digitalEdgeTriggerSource;
        RFmxCdma2kMXDigitalEdgeTriggerEdge digitalEdgeTriggerEdge;
        double triggerDelay;
        bool enableTrigger;

        double carrierAbsolutePower, timeout;

        internal void Run()
        {
            try
            {
                InitializeVariables();
                InitializeInstr();
                ConfigureCdma2k();
                RetrieveResults();
                PrintResults();

            }
            catch (Exception ex)
            {
                DisplayError(ex);
            }
            finally
            {
                /* Close session */
                CloseSession();
                Console.WriteLine("Press any key to exit.....");
                Console.ReadKey();
            }
        }

        private void InitializeVariables()
        {
            /* Initialize input variables */

            resourceName = "RFSA";
            centerFrequency = 833.49e+6;                                        /* Hz */
            referenceLevel = 0.00;                                              /* dBm */
            externalAttenuation = 0.00;                                         /* dB */
            timeout = 10.0;                                                     /* seconds */

            frequencySource = RFmxInstrMXConstants.OnboardClock;
            frequency = 10.0e+6;                                                /* Hz */

            //Sweep time
            sweepTimeAuto = RFmxCdma2kMXChpSweepTimeAuto.True;
            sweepTimeInterval = 1.67e-3;                                        /* seconds */

            //Averaging
            averagingEnabled = RFmxCdma2kMXChpAveragingEnabled.False;
            averagingCount = 10;
            averagingType = RFmxCdma2kMXChpAveragingType.Rms;

            triggerDelay = 0.00;                                                /* seconds */

            digitalEdgeTriggerSource = RFmxInstrMXConstants.Pfi0;
            digitalEdgeTriggerEdge = RFmxCdma2kMXDigitalEdgeTriggerEdge.Rising;
            enableTrigger = false;
        }

        private void InitializeInstr()
        {
            /* Create a new RFmx Session */
            instrSession = new RFmxInstrMX(resourceName, "");
        }

        private void ConfigureCdma2k()
        {
            /* Get SpecAn signal */
            cdma2k = instrSession.GetCdma2kSignalConfiguration();

            /* Configure measurement */
            instrSession.ConfigureFrequencyReference("", frequencySource, frequency);
            cdma2k.ConfigureRF("", centerFrequency, referenceLevel, externalAttenuation);
            cdma2k.ConfigureDigitalEdgeTrigger("", digitalEdgeTriggerSource, digitalEdgeTriggerEdge, triggerDelay, enableTrigger);

            cdma2k.SelectMeasurements("", RFmxCdma2kMXMeasurementTypes.Chp, true);

            cdma2k.Chp.Configuration.ConfigureSweepTime("", sweepTimeAuto, sweepTimeInterval);
            cdma2k.Chp.Configuration.ConfigureAveraging("", averagingEnabled, averagingCount, averagingType);

            cdma2k.Initiate("", "");
        }

        private void RetrieveResults()
        {
            /* Retrieve results */
            Spectrum<float> spectrum = null;
            cdma2k.Chp.Results.FetchCarrierAbsolutePower("", timeout, out carrierAbsolutePower);
            cdma2k.Chp.Results.FetchSpectrum("", timeout, ref spectrum);            
        }

        private void PrintResults()
        {
            Console.WriteLine("Carrier Absolute Power (dBm)  : {0}", carrierAbsolutePower);
        }

        private void CloseSession()
        {
            try
            {
                if (cdma2k != null)
                {
                    cdma2k.Dispose();
                    cdma2k = null;
                }

                if (instrSession != null)
                {
                    instrSession.Close();
                    instrSession = null;
                }
            }
            catch (Exception ex)
            {
                DisplayError(ex);
            }
        }

        static private void DisplayError(Exception ex)
        {
            Console.WriteLine("ERROR:\n" + ex.GetType() + ": " + ex.Message);
        }
    }
}
